### PR TITLE
WFP ALE Flow established callout, tests etc.

### LINKED ISSUE #786: Add sock ops hook.
Documentation at section 1.2 of https://blogs.oracle.com/linux/post/bpf-a-tour-of-program-types.

Programs of this type can intercept `"socket operations such as connection establishment, retransmit timeout etc. Once caught options can also be set via bpf_setsockopt()..."`

This can be useful for building network observability solutions using eBPF for Windows.

- [ ] Program type: `BPF_PROG_TYPE_SOCK_OPS`
- [ ] Attach type: `BPF_CGROUP_SOCK_OPS`

The following sock operation types should be supported.
- [ ] BPF_SOCK_OPS_TCP_CONNECT_CB,	/* Calls BPF program right before an
					 * active connection is initialized
					 */
- [ ] BPF_SOCK_OPS_ACTIVE_ESTABLISHED_CB,	/* Calls BPF program when an
						 * active connection is
						 * established
						 */
- [ ] BPF_SOCK_OPS_PASSIVE_ESTABLISHED_CB,	/* Calls BPF program when a
						 * passive connection is
						 * established
						 
- [ ]    BPF_SOCK_OPS_CONNECTION_DELETED_CB,    ///< Calls BPF program when a
                         ///< connection is deleted.

### PR DESCRIPTION
**Pull request by shankarseal**
## Description

Fixes #786 

- WFP callout for WFP ALE established. The callout invokes the eBPF program notifying about new inbound/outbound connections. If the eBPF program returns success, the WFP flow is associated with a context.
- Upon flow deletion, eBPF program is notified of connection tearing down.
- Some changes to common code in netebpfext.
- Fixed some race conditions between WFP callbacks and program detach.
- Changes to socket and ring-buffer test libraries.
- Ran generate_expected_bpf2c_output.ps1.
- Some minor CICD WF changes.

## Testing

New test cases added in socket_tests.exe

## Documentation

None.

### CODE REVIEW

File: netebpfext/net_ebpf_ext_sock_ops.c [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+}
+
+void
+net_ebpf_extension_sock_ops_flow_delete(uint16_t layer_id, uint32_t callout_id, uint64_t flow_context)
+{
+    net_ebpf_extension_sock_ops_wfp_flow_context_t* local_flow_context =
+        (net_ebpf_extension_sock_ops_wfp_flow_context_t*)(uintptr_t)flow_context;
+    net_ebpf_extension_sock_ops_wfp_filter_context_t* filter_context;
+    net_ebpf_extension_hook_client_t* attached_client = NULL;
+    bpf_sock_ops_t* sock_ops_ctx = NULL;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_sock_ops.c:414]:
```suggestion
    bpf_sock_ops_t* sock_ops_context = NULL;
```

File: netebpfext/net_ebpf_ext.c [hunk 2]
```diff
@@ -156,6 +189,72 @@ static HANDLE _fwp_engine_handle;
...
+    net_ebpf_extension_hook_id_t hook_id = 0;
+    ASSERT(
+        (wfp_layer_id == FWPS_LAYER_OUTBOUND_MAC_FRAME_NATIVE) ||
+        (wfp_layer_id == FWPS_LAYER_INBOUND_MAC_FRAME_NATIVE) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_RESOURCE_ASSIGNMENT_V4) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_RESOURCE_ASSIGNMENT_V6) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_RESOURCE_RELEASE_V4) || (wfp_layer_id == FWPS_LAYER_ALE_RESOURCE_RELEASE_V6) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_AUTH_CONNECT_V4) || (wfp_layer_id == FWPS_LAYER_ALE_AUTH_CONNECT_V6) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_AUTH_RECV_ACCEPT_V4) || (wfp_layer_id == FWPS_LAYER_ALE_AUTH_RECV_ACCEPT_V6) ||
+        (wfp_layer_id == FWPS_LAYER_ALE_FLOW_ESTABLISHED_V4) || (wfp_layer_id == FWPS_LAYER_ALE_FLOW_ESTABLISHED_V6));
```
[the remark this answers is not in the document]
Comment (shankarseal) [reply netebpfext/net_ebpf_ext.c:204]: The reason why I added this seemingly-ridiculous assert is because one of the previous feedback was to avoid ASSERT(FALSE). I will gladly change this.

File: .github/workflows/cicd.yml [hunk 3]
```diff
@@ -64,18 +64,6 @@ jobs:
       code_coverage: true
       gather_dumps: true
 
-  # Run the API tests in GitHub.
```
Comment (Alan-Jowett) [inline .github/workflows/cicd.yml:67]: Please move this to a separate PR. this PR is big enough as is.

File: netebpfext/net_ebpf_ext.h [hunk 4]
```diff
@@ -60,6 +73,98 @@ typedef struct _net_ebpf_extension_wfp_filter_parameters
...
 } net_ebpf_extension_wfp_filter_parameters_t;
 
+/**
+ * "Base class" for all WFP filter contexts used by net ebpf extension hooks.
+ */
+typedef struct _net_ebpf_extension_wfp_filter_context
+{
+    volatile long reference_count;                                ///< Reference count.
+    const struct _net_ebpf_extension_hook_client* client_context; ///< Pointer to hook NPI client.
+    uint64_t* filter_ids;                                         // Array of WFP filter Ids.
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext.h:83]:
```suggestion
    uint64_t* filter_ids;                                         ///< Array of WFP filter Ids.
```

File: netebpfext/net_ebpf_ext_xdp.c [hunk 5]
```diff
@@ -51,9 +51,8 @@ const net_ebpf_extension_wfp_filter_parameters_t _net_ebpf_extension_xdp_wfp_fil
 
 typedef struct _net_ebpf_extension_xdp_wfp_filter_context
 {
-    const net_ebpf_extension_hook_client_t* client_context;
+    net_ebpf_extension_wfp_filter_context_t;
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_xdp.c:54]: not a big deal, but I prefer always using named fields so you don't need to have type casts, but can use CONTAINING_RECORD() to go from net_ebpf_extension_wfp_filter_context_t to net_ebpf_extension_xdp_wfp_filter_context_t, and just use the field by name to go in the other direction.  This is not blocking, I will approve either way but wanted to pass on what I think is best practice in coding conventions, since it's much more typesafe than using casts which can be bug prone.

File: netebpfext/net_ebpf_ext_sock_ops.c [hunk 6]
```diff
@@ -0,0 +1,487 @@
...
+#include "net_ebpf_ext_sock_ops.h"
+
+//
+// WFP related types & globals for SOCK_OPS hook.
+//
+
+struct _net_ebpf_extension_sock_ops_wfp_filter_context;
+
+/**
+ * @brief Custom context associated with WFP flows that are notified to eBPF programs
```
Comment (dthaler) [inline netebpfext/net_ebpf_ext_sock_ops.c:21]:
```suggestion
 * @brief Custom context associated with WFP flows that are notified to eBPF programs.
```

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
